plan.lua

```diff
@@ -213,100 +213,174 @@ function plan_class:get_random_plan_pos()
 	end
 end
 
+
 --------------------------------------
--- Generate a plan from schematics file
+-- return file extension name or nil from filename
 --------------------------------------
--- replacements parameter is optional, and only used by .mts files.
-function plan_class:read_from_schem_file(filename, replacements)
+local function get_ext_name(filename)
+	return string.match(filename, '^.+(%..+)$')
+end
 
-	local mts_format = string.find(filename, '.mts',  -4)
+--------------------------------------
+-- add MTS Object to nodes at optional position(defaults to {x=0,y=0,z=0})
+--------------------------------------
+function plan_class:add_mts_obj(mts, pos, replacements)
+	local size = mts.size
+	local size_x = size.x
+	local size_y = size.y
+	local size_z = size.z
+	if type(pos) ~= "table" then pos = {} end
+	local ox = pos['x'] or 0
+	local oy = pos['y'] or 0
+	local oz = pos['z'] or 0
+
+	-- read node names
+	local node_name_count = mts.node_name_count
+	local node_names = mts.node_names
+	local air_id = 0;
+	for i = 1, node_name_count do
+		local name_text   = node_names[i]
+		if replacements and replacements[name_text] then
+			name_text = replacements[name_text]
+			node_names[i] = name_text
+		end
+		if name_text == "air" then
+			air_id = i
+		end
+	end
 
-	local readingMode = 'r'
-	if mts_format then 
-		-- Minetest Schematics are a packed binary file format
-		readingMode = 'rb'
+	local compressed_data = mts.data
+	local data_string = minetest.decompress(compressed_data, "deflate" );
+
+	local i = 1
+	local p2offset = (size_x*size_y*size_z)*3
+	for z = 1, size_z do
+		for y = 1, size_y do
+			for x = 1, size_x do
+				local id = string.byte(data_string, i) * 256 + string.byte(data_string, i+1)
+				i = i + 2
+				local p2 = string.byte(data_string, p2offset + math.floor(i/2))
+				id = id + 1
+				if id ~= air_id then
+					local npos = {x = x+ox, y = y+oy, z = z+oz}
+					local ent = {name = node_names[id], param2 = p2}
+					self:add_node(npos, ent)
+					self:adjust_building_info(npos, ent)
+				end
+			end
+		end
 	end
+end
+
 
-	local file = io.open( filename, readingMode )
+--------------------------------------
+-- add WorldEdit Object to nodes at optional position(defaults to {x=0,y=0,z=0})
+--------------------------------------
+function plan_class:add_we_obj(we, pos, replacements)
+	if type(pos) ~= "table" then pos = {} end
+	local ox = pos['x'] or 0
```

Comment:
Not importent, but the usual syntax in lua is `pos.x`instead of `pos['x']`

Reply:
I just assign it as empty table if pos is nil. So the pos.x will throw error.

Reply:
Okay, I admit I think too much.